homeassistant/components/generic_thermostat/climate.py

```diff
@@ -314,12 +372,12 @@ def hvac_modes(self):
     @property
     def preset_mode(self):
         """Return the current preset mode, e.g., home, away, temp."""
-        return PRESET_AWAY if self._is_away else PRESET_NONE
+        return self._current_preset
 
     @property
     def preset_modes(self):
-        """Return a list of available preset modes or PRESET_NONE if _away_temp is undefined."""
-        return [PRESET_NONE, PRESET_AWAY] if self._away_temp else PRESET_NONE
+        """Return a list of available preset modes or PRESET_NONE if _presets is undefined."""
+        return list(self._presets.keys()) if self._presets else PRESET_NONE
```

Comment:
Keys is the default iterator when iterating a dict. If there are no presets, we should not return a string value. This should be a list or None. 

```suggestion
        return list(self._presets) if self._presets else None
```